File: src/main/java/de/istkorrekt/gameoflife/Cell.java
package de.istkorrekt.gameoflife;

public class Cell {

    private static final Cell CANONICAL_DEAD = new Cell(State.DEAD);
    private static final Cell CANONICAL_ALIVE = new Cell(State.ALIVE);

    public enum State {
        DEAD, ALIVE
    }

    private final State state;

    private Cell(State state) {
        this.state = state;
    }

    public Cell nextGeneration(long livingNeighbors) {
        if (isAlive()) {
            if (livingNeighbors < 2) {
                return dead();
            }
            if (livingNeighbors < 4) {
                return alive();
            }
            return dead();
        }
        if (livingNeighbors == 3) {
            return alive();
        }
        return dead();
    }

    public boolean isAlive() {
        return state == State.ALIVE;
    }

    public static Cell alive() {
        return CANONICAL_ALIVE;
    }

    public static Cell dead() {
        return CANONICAL_DEAD;
    }

    @Override
    public String toString() {
        switch (state) {
            case DEAD:
                return "[ ]";
            case ALIVE:
                return "[❤]";
        }
        throw new IllegalStateException();
    }
}

File: src/main/kotlin/de/istkorrekt/gameoflife/Cell.kt
package de.istkorrekt.gameoflife

enum class Cell {
    DEAD,
    ALIVE;

    fun nextGeneration(livingNeighbors: Int): Cell {
        if (this == ALIVE) {
            if (livingNeighbors < 2L) {
                return DEAD
            }
            return if (livingNeighbors < 4L) {
                ALIVE
            } else DEAD
        }
        return if (livingNeighbors == 3) {
            ALIVE
        } else DEAD
    }

    fun isAlive(): Boolean = this == ALIVE

    override fun toString(): String {
        return when (this) {
            DEAD -> "[ ]"
            ALIVE -> "[❤]"
        }
    }
}